File: server/agent/parsers.ts
import type {
  ConstraintDelta,
  FeedbackDimension,
  ParsedNeed
} from "./contracts.js";
import { type FeedbackPayloadInput } from "./types.js";

const colorLexicon: Record<string, string> = {
  黑: "black",
  白: "white",
  红: "red",
  蓝: "blue",
  绿: "green",
  黄: "yellow",
  粉: "pink",
  紫: "purple",
  灰: "gray",
  米: "beige",
  navy: "navy",
  black: "black",
  white: "white",
  red: "red",
  blue: "blue",
  beige: "beige",
  green: "green",
  yellow: "yellow",
  pink: "pink",
  purple: "purple",
  gray: "gray",
  grey: "gray",
  brown: "brown",
  khaki: "khaki",
  cream: "cream"
};

const categoryLexicon: Record<string, string> = {
  外套: "outerwear",
  夹克: "outerwear",
  上衣: "top",
  衬衫: "top",
  毛衣: "top",
  裤: "bottom",
  裙: "bottom",
  连衣裙: "dress",
  鞋: "shoes",
  jacket: "outerwear",
  coat: "outerwear",
  blazer: "outerwear",
  shirt: "top",
  top: "top",
  "t-shirt": "top",
  tshirt: "top",
  sweater: "top",
  blouse: "top",
  hoodie: "top",
  pants: "bottom",
  jeans: "bottom",
  trousers: "bottom",
  shorts: "bottom",
  skirt: "bottom",
  dress: "dress",
  shoes: "shoes",
  sneakers: "shoes",
  boots: "shoes",
  heels: "shoes"
};

const styleLexicon: Record<string, string> = {
  休闲: "casual",
  通勤: "business",
  正式: "formal",
  简约: "minimal",
  街头: "streetwear",
  运动: "sporty",
  可爱: "romantic",
  casual: "casual",
  business: "business",
  formal: "formal",
  minimal: "minimal",
  minimalist: "minimal",
  street: "streetwear",
  sporty: "sporty",
  athletic: "sporty",
  cute: "romantic",
  dressy: "formal",
  elegant: "classic"
};

// Coarse formality the customer asks for → the product `formality` facet.
const formalityLexicon: Record<string, string> = {
  正式: "formal",
  正装: "formal",
  商务: "formal",
  通勤: "formal",
  庄重: "formal",
  隆重: "formal",
  formal: "formal",
  dressy: "formal",
  商务休闲: "smart_casual",
  轻正式: "smart_casual",
  "smart casual": "smart_casual",
  "smart-casual": "smart_casual",
  休闲: "casual",
  随意: "casual",
  日常: "casual",
  居家: "casual",
  casual: "casual",
  everyday: "casual",
  relaxed: "casual"
};

// Colour tone the customer asks for → the product `color_tone` facet.
const toneLexicon: Record<string, string> = {
  深色: "dark",
  深: "dark",
  暗色: "dark",
  暗: "dark",
  "深色系": "dark",
  dark: "dark",
  darker: "dark",
  浅色: "light",
  浅: "light",
  淡色: "light",
  亮色: "light",
  "浅色系": "light",
  light: "light",
  lighter: "light",
  pale: "light",
  pastel: "light"
};

export function routeIntent(text: string): "explicit" | "recommendation" | "unclear" {
  const normalized = text.toLowerCase();
  // Concrete preference words → treat as an explicit styling request.
  if (Object.keys(categoryLexicon).some((keyword) => normalized.includes(keyword))) return "explicit";
  if (Object.keys(styleLexicon).some((keyword) => normalized.includes(keyword))) return "explicit";
  if (Object.keys(colorLexicon).some((keyword) => normalized.includes(keyword))) return "explicit";
  if (Object.keys(formalityLexicon).some((keyword) => normalized.includes(keyword))) return "explicit";
  if (Object.keys(toneLexicon).some((keyword) => normalized.includes(keyword))) return "explicit";
  // Occasion words imply a concrete need even when no garment is named.
  if (/约会|婚礼|婚宴|面试|上班|通勤|派对|聚会|宴会|旅行|旅游|度假|海边|运动|健身|约会|正式场合|商务|开会/.test(normalized))
    return "explicit";
  if (/\b(wedding|interview|date|party|work|office|trip|travel|vacation|beach|gym|dinner|meeting)\b/.test(normalized))
    return "explicit";
  if (/预算|以内|日元|yen|想要|找/.test(normalized)) return "explicit";
  if (/budget|under\s*[¥$]?\s*\d|within\s*[¥$]?\s*\d|\$|dollar|cheap/.test(normalized)) return "explicit";
  // "帮我挑几套 / 帮我搭 / 搭配 / 穿搭 / 来几套 / 推荐" → let the agent suggest.
  if (
    /没.*想法|不知道|推荐|随便|适合我|你来|帮我挑|挑[几一]套|挑套|帮我搭|搭[几一]套|搭配|穿搭|来[几一]套|看看有什么/.test(
      normalized
    )
  )
    return "recommendation";
  // "recommend something / pick for me / style me / an outfit / a look" → let the agent suggest.
  if (
    /recommend|suggest|no idea|not sure|don'?t know|whatever|surprise me|up to you|pick (?:for me|something|a few|some)|help me (?:pick|choose|find)|style me|put together|outfit|a look|some looks|match (?:my|what)/.test(
      normalized
    )
  )
    return "recommendation";
  return "unclear";
}

export function parseNeed(text: string): ParsedNeed {
  const colors = collectMatches(text, colorLexicon);
  const categories = collectMatches(text, categoryLexicon);
  const styleTags = collectMatches(text, styleLexicon);
  const budget = parseBudgetYen(text);
  return {
    raw_text: text,
    categories,
    colors,
    style_tags: styleTags,
    budget_yen: budget,
    formality: collectMatches(text, formalityLexicon)[0],
    color_tone: collectMatches(text, toneLexicon)[0]
  };
}

export function parseFeedback(feedback: FeedbackPayloadInput): ConstraintDelta {
  if (feedback.feedback_type === "confirm") {
    return {
      prefer: [],
      avoid: [],
      requires_new_recommendation: false,
      notes: "confirmed"
    };
  }

  if (feedback.feedback_type === "positive_keep") {
    const dimension = feedback.dimension ?? "overall";
    return {
      prefer: [
        {
          dimension,
          value: feedback.dimension_value ?? "current_selection",
          reason: "positive_feedback"
        }
      ],
      avoid: [],
      requires_new_recommendation: false,
      notes: "positive_feedback"
    };
  }

  if (feedback.feedback_type === "reject_all") {
    return {
      prefer: [],
      avoid: [
        {
          dimension: "overall",
          value: feedback.set_id,
          reason: "reject_all"
        }
      ],
      requires_new_recommendation: true,
      notes: "reject_all"
    };
  }

  const text = feedback.raw_voice_text ?? "";
  const dimension = feedback.dimension ?? inferFeedbackDimension(text);
  const value = feedback.dimension_value ?? inferFeedbackValue(dimension, text) ?? "unspecified";
  const budget = dimension === "price" ? parseBudgetYen(text) : undefined;

  return {
    prefer: buildPreference(dimension, value),
    avoid: [
      {
        dimension,
        value,
        reason: "negative_feedback"
      }
    ],
    budget_yen: budget,
    requires_new_recommendation: true,
    notes: `${dimension}_feedback`
  };
}

function buildPreference(dimension: FeedbackDimension, value: string) {
  if (dimension === "style" && /正式|formal|too_formal/.test(value)) {
    return [
      {
        dimension,
        value: "casual",
        reason: "style_counter_preference"
      }
    ];
  }
  if (dimension === "price") {
    return [
      {
        dimension,
        value: "lower_price",
        reason: "price_counter_preference"
      }
    ];
  }
  return [];
}

function collectMatches(text: string, lexicon: Record<string, string>) {
  const normalized = text.toLowerCase();
  return Array.from(
    new Set(
      Object.entries(lexicon)
        .filter(([keyword]) => normalized.includes(keyword.toLowerCase()))
        .map(([, value]) => value)
    )
  );
}

function parseBudgetYen(text: string) {
  const match = /(\d+(?:\.\d+)?)\s*(万|w|千|k|円|日元|yen)?/i.exec(text);
  if (!match) return undefined;

  const value = Number(match[1]);
  const unit = match[2]?.toLowerCase();
  if (Number.isNaN(value)) return undefined;
  if (unit === "万" || unit === "w") return Math.round(value * 10000);
  if (unit === "千" || unit === "k") return Math.round(value * 1000);
  return Math.round(value);
}

function inferFeedbackDimension(text: string): FeedbackDimension {
  if (/颜色|色|colou?r|red|blue|black|white/i.test(text)) return "color";
  if (/版型|剪裁|宽|紧|fit|size|loose|tight|baggy/i.test(text)) return "fit";
  if (/风格|正式|休闲|style|formal|casual|dressy/i.test(text)) return "style";
  if (/贵|便宜|预算|price|expensive|cheap|pricey|cost|budget/i.test(text)) return "price";
  return "overall";
}

function inferFeedbackValue(dimension: FeedbackDimension, text: string) {
  if (dimension === "color") return collectMatches(text, colorLexicon)[0];
  if (dimension === "style") {
    if (/太正式|too formal|formal/i.test(text)) return "too_formal";
    return collectMatches(text, styleLexicon)[0];
  }
  if (dimension === "price") return parseBudgetYen(text)?.toString();
  if (dimension === "fit") {
    if (/宽|loose|oversized|baggy/i.test(text)) return "too_loose";
    if (/紧|tight/i.test(text)) return "too_tight";
  }
  return undefined;
}

File: agent_foundation/parsers.py
"""Rule-based parsers for the first Agent foundation prototype.

The first implementation is deliberately deterministic so tests and demos are
stable. Later, these functions can become Gemini/ADK tool-backed adapters.
"""

from __future__ import annotations

import re

from .contracts import (
    BudgetRange,
    ConstraintDelta,
    ConstraintItem,
    FeedbackDimension,
    FeedbackPayload,
    FeedbackType,
    Route,
    UserNeedConstraints,
)


RECOMMENDATION_PATTERNS = (
    "没想法",
    "沒有想法",
    "不知道",
    "你推荐",
    "你推薦",
    "推荐适合",
    "推薦適合",
    "帮我搭",
    "幫我搭",
    "随便",
    "適合我",
    "适合我",
)

EXPLICIT_PATTERNS = (
    "预算",
    "預算",
    "日元",
    "円",
    "外套",
    "上衣",
    "裤",
    "褲",
    "裙",
    "鞋",
    "约会",
    "約會",
    "面试",
    "面試",
    "通勤",
    "海边",
    "海邊",
    "黑",
    "白",
    "红",
    "紅",
    "蓝",
    "藍",
    "休闲",
    "休閒",
    "正式",
    "街头",
    "街頭",
)


def route_intent(text: str) -> Route:
    normalized = text.strip().lower()
    if not normalized:
        return Route.UNCLEAR

    if any(pattern in normalized for pattern in RECOMMENDATION_PATTERNS):
        if not any(pattern in normalized for pattern in EXPLICIT_PATTERNS):
            return Route.RECOMMENDATION

    if any(pattern in normalized for pattern in EXPLICIT_PATTERNS):
        return Route.EXPLICIT

    if len(normalized) <= 8:
        return Route.UNCLEAR
    return Route.UNCLEAR


def parse_need(text: str) -> UserNeedConstraints:
    need = UserNeedConstraints(free_text_summary=text.strip() or None)
    lower = text.lower()

    category_map = {
        "outerwear": ("外套", "夹克", "jacket", "coat"),
        "top": ("上衣", "内搭", "衬衫", "shirt", "top"),
        "bottom": ("裤", "褲", "长裤", "短裤", "pants"),
        "dress": ("裙", "连衣裙", "dress"),
        "shoes": ("鞋", "sneaker", "boots"),
    }
    for category, patterns in category_map.items():
        if any(pattern in lower for pattern in patterns):
            need.category.append(category)

    style_map = {
        "minimal": ("简约", "簡約", "minimal"),
        "casual": ("休闲", "休閒", "casual", "不太正式"),
        "formal": ("正式", "面试", "面試", "formal"),
        "street": ("街头", "街頭", "street"),
        "date": ("约会", "約會", "date"),
        "vacation": ("海边", "海邊", "旅行", "vacation"),
        "office": ("通勤", "office"),
    }
    for style, patterns in style_map.items():
        if any(pattern in lower for pattern in patterns):
            need.style_tags.append(style)

    color_map = {
        "black": ("黑", "black"),
        "white": ("白", "white"),
        "red": ("红", "紅", "red"),
        "blue": ("蓝", "藍", "blue"),
        "beige": ("米色", "beige"),
    }
    for color, patterns in color_map.items():
        if any(pattern in lower for pattern in patterns):
            need.colors.append(color)

    budget = _parse_budget_yen(lower)
    if budget:
        need.budget_range = BudgetRange(min=0, max=budget, currency="JPY")

    occasion_map = {
        "beach": ("海边", "海邊"),
        "date": ("约会", "約會"),
        "interview": ("面试", "面試"),
        "commute": ("通勤",),
    }
    for occasion, patterns in occasion_map.items():
        if any(pattern in lower for pattern in patterns):
            need.occasion = occasion
            break

    if "不喜欢露腿" in lower or "不想露腿" in lower:
        need.avoid.append(
            ConstraintItem(
                dimension=FeedbackDimension.FIT.value,
                value="leg_exposure",
                reason="customer_disliked",
            )
        )

    return _dedupe_need(need)


def parse_feedback(feedback: FeedbackPayload) -> ConstraintDelta:
    if feedback.feedback_type == FeedbackType.CONFIRM:
        return ConstraintDelta(requires_new_recommendation=False)

    if feedback.feedback_type == FeedbackType.POSITIVE_KEEP:
        return ConstraintDelta(
            avoid_add=[],
            keep_add=[
                ConstraintItem(
                    dimension=(feedback.dimension or FeedbackDimension.OVERALL).value,
                    value=feedback.dimension_value or "current_selection",
                    reason="positive_feedback",
                )
            ],
            requires_new_recommendation=True,
        )

    if feedback.feedback_type == FeedbackType.REJECT_ALL:
        return ConstraintDelta(
            avoid_add=[
                ConstraintItem(
                    dimension=FeedbackDimension.OVERALL.value,
                    value=feedback.set_id,
                    reason="reject_all",
                )
            ],
            requires_new_recommendation=True,
            notes="broaden_candidate_pool",
        )

    dimension = feedback.dimension or _infer_feedback_dimension(
        feedback.raw_voice_text or ""
    )
    value = feedback.dimension_value or _infer_feedback_value(
        dimension, feedback.raw_voice_text or ""
    )

    if dimension == FeedbackDimension.PRICE:
        budget = _parse_budget_yen(feedback.raw_voice_text or "")
        return ConstraintDelta(
            budget_update=BudgetRange(min=0, max=budget, currency="JPY")
            if budget
            else None,
            requires_new_recommendation=True,
            notes="price_feedback",
        )

    if dimension == FeedbackDimension.STYLE:
        return ConstraintDelta(
            style_shift=value,
            requires_new_recommendation=True,
            notes="style_feedback",
        )

    if dimension == FeedbackDimension.FIT:
        return ConstraintDelta(
            avoid_add=[
                ConstraintItem(
                    dimension=dimension.value,
                    value=value or "current_fit",
                    reason="customer_disliked",
                )
            ],
            fit_update=value,
            requires_new_recommendation=True,
        )

    return ConstraintDelta(
        avoid_add=[
            ConstraintItem(
                dimension=dimension.value,
                value=value or "current_selection",
                reason="customer_disliked",
            )
        ],
        requires_new_recommendation=True,
    )


def _parse_budget_yen(text: str) -> int | None:
    # Covers "两万日元", "2万日元", "20000円", and "一万以内" style inputs.
    cn_numbers = {
        "一": 1,
        "二": 2,
        "两": 2,
        "兩": 2,
        "三": 3,
        "四": 4,
        "五": 5,
        "六": 6,
        "七": 7,
        "八": 8,
        "九": 9,
    }
    match = re.search(r"(\d+)\s*万", text)
    if match:
        return int(match.group(1)) * 10000
    for char, number in cn_numbers.items():
        if f"{char}万" in text:
            return number * 10000
    match = re.search(r"(\d{4,6})\s*(日元|円|yen|jpy)?", text)
    if match and ("预算" in text or "預算" in text or "以内" in text or "以內" in text):
        return int(match.group(1))
    return None


def _infer_feedback_dimension(text: str) -> FeedbackDimension:
    lower = text.lower()
    if any(token in lower for token in ("颜色", "顏色", "色", "太亮", "太暗")):
        return FeedbackDimension.COLOR
    if any(token in lower for token in ("版型", "剪裁", "修身", "宽松", "寬鬆")):
        return FeedbackDimension.FIT
    if any(token in lower for token in ("风格", "風格", "正式", "休闲", "休閒")):
        return FeedbackDimension.STYLE
    if any(token in lower for token in ("贵", "貴", "便宜", "预算", "預算")):
        return FeedbackDimension.PRICE
    return FeedbackDimension.OVERALL


def _infer_feedback_value(dimension: FeedbackDimension, text: str) -> str | None:
    lower = text.lower()
    if dimension == FeedbackDimension.COLOR:
        color_map = {
            "red": ("红", "紅", "red"),
            "black": ("黑", "black"),
            "white": ("白", "white"),
            "blue": ("蓝", "藍", "blue"),
            "bright": ("太亮", "亮"),
            "dark": ("太暗", "暗"),
        }
        for value, patterns in color_map.items():
            if any(pattern in lower for pattern in patterns):
                return value
    if dimension == FeedbackDimension.FIT:
        if "修身" in lower:
            return "slim"
        if "宽松" in lower or "寬鬆" in lower:
            return "loose"
        if "露腿" in lower:
            return "leg_exposure"
    if dimension == FeedbackDimension.STYLE:
        if "正式" in lower:
            return "formal"
        if "休闲" in lower or "休閒" in lower:
            return "casual"
    return None


def _dedupe_need(need: UserNeedConstraints) -> UserNeedConstraints:
    need.category = _dedupe(need.category)
    need.style_tags = _dedupe(need.style_tags)
    need.colors = _dedupe(need.colors)
    need.avoid = _dedupe_constraints(need.avoid)
    need.keep = _dedupe_constraints(need.keep)
    return need


def _dedupe(values: list[str]) -> list[str]:
    result: list[str] = []
    for value in values:
        if value not in result:
            result.append(value)
    return result


def _dedupe_constraints(items: list[ConstraintItem]) -> list[ConstraintItem]:
    result: list[ConstraintItem] = []
    seen: set[tuple[str, str]] = set()
    for item in items:
        key = (item.dimension, item.value)
        if key not in seen:
            result.append(item)
            seen.add(key)
    return result
